applications/RomApplication/custom_strategies/petrov_galerkin_rom_builder_and_solver.h

```diff
@@ -0,0 +1,541 @@
+
+/* Application includes */
+#include "rom_application_variables.h"
+#include "custom_utilities/rom_auxiliary_utilities.h"
+
+namespace Kratos
+{
+
+///@name Kratos Globals
+///@{
+
+
+///@}
+///@name Type Definitions
+///@{
+
+
+///@}
+///@name  Enum's
+///@{
+
+
+///@}
+///@name  Functions
+///@{
+
+
+///@}
+///@name Kratos Classes
+///@{
+
+template <class TSparseSpace, class TDenseSpace, class TLinearSolver>
+class PetrovGalerkinROMBuilderAndSolver : public ROMBuilderAndSolver<TSparseSpace, TDenseSpace, TLinearSolver>
+{
+public:
+
+    //TODO: UPDATE THIS
+    /**
+     * This struct is used in the component wise calculation only
+     * is defined here and is used to declare a member variable in the component wise builder and solver
+     * private pointers can only be accessed by means of set and get functions
+     * this allows to set and not copy the Element_Variables and Condition_Variables
+     * which will be asked and set by another strategy object
+     */
+
+    ///@name Type Definitions
+    ///@{
+
+    // Class pointer definition
+    KRATOS_CLASS_POINTER_DEFINITION(PetrovGalerkinROMBuilderAndSolver);
+
+    // The size_t types
+    typedef std::size_t SizeType;
+    typedef std::size_t IndexType;
+
+    /// The definition of the current class
+    typedef PetrovGalerkinROMBuilderAndSolver<TSparseSpace, TDenseSpace, TLinearSolver> ClassType;
+
+    /// Definition of the classes from the base class
+    typedef ROMBuilderAndSolver<TSparseSpace, TDenseSpace, TLinearSolver> BaseType;
+    typedef typename BaseType::TSchemeType TSchemeType;
+    typedef typename BaseType::DofsArrayType DofsArrayType;
+    typedef typename BaseType::TSystemMatrixType TSystemMatrixType;
+    typedef typename BaseType::TSystemVectorType TSystemVectorType;
+    typedef typename BaseType::LocalSystemVectorType LocalSystemVectorType;
+    typedef typename BaseType::LocalSystemMatrixType LocalSystemMatrixType;
+    typedef typename BaseType::TSystemMatrixPointerType TSystemMatrixPointerType;
+    typedef typename BaseType::TSystemVectorPointerType TSystemVectorPointerType;
+    typedef typename BaseType::ElementsArrayType ElementsArrayType;
+    typedef typename BaseType::ConditionsArrayType ConditionsArrayType;
+
+    /// Additional definitions
+    typedef typename ModelPart::MasterSlaveConstraintContainerType MasterSlaveConstraintContainerType;
+    typedef Element::EquationIdVectorType EquationIdVectorType;
+    typedef Element::DofsVectorType DofsVectorType;
+    typedef boost::numeric::ublas::compressed_matrix<double> CompressedMatrixType;
+
+    // Non-distributed, dense:
+    typedef LocalSystemMatrixType RomSystemMatrixType;
+    typedef LocalSystemVectorType RomSystemVectorType;
+
+    //Distributed, dense
+    typedef RomSystemMatrixType PetrovGalerkinSystemMatrixType; 
+    typedef RomSystemVectorType PetrovGalerkinSystemVectorType;
+    //      ^ Change this to a distributed dense type
+

```

Comment:
Are dense vector `PetrovGalerkinSystemVectorType` and dense matrix `PetrovGalerkinSystemMatrixType` distributed or not?

Reply:
Not yet, I will change it in a future PR.